allow step function to use other lambda functions as task

mg-epsy opened the pull request:
1. currently, only functions defined in the same serverless.yml is allowed as tasks. This change is to allow step function to use other lambda functions as task.
2. Removing `End` since Parallel `End` property causes error during sls deploy.
https://www.serverless.com/plugins/serverless-step-functions#parallel

```
stepFunctions:
  stateMachines:
    yourParallelMachine:
      definition:
        Comment: "An example of the Amazon States Language using a parallel state to execute two branches at the same time."
        StartAt: Parallel
        States:
          Parallel:
            Type: Parallel
            Next: Final State
            Branches:
            - StartAt: Wait 20s
              States:
                Wait 20s:
                  Type: Wait
                  Seconds: 20
                  End: true
            - StartAt: Pass
              States:
                Pass:
                  Type: Pass
                  Next: Wait 10s
                Wait 10s:
                  Type: Wait
                  Seconds: 10
                  End: true
          Final State:
            Type: Pass
            End: true
```

mg-epsy committed:
allow step function to use other lambda functions as task

allow step function to use other lambda functions as task

whisller commented on serverless/aws/features/stepfunctions.py:
```diff
@@ -337,8 +337,15 @@ def __init__(
-    def task(self, function, end: Optional[bool] = None):
-        return self.definition.add(Task(function, end=end))
+    
+    def task(self, function=None, resource=None, name=None, end: Optional[bool] = None):
```
Wouldn't this code behave this same way: 
```python
    def task(self, function=None, resource=None, name=None, end: Optional[bool] = None):
        return self.definition.add(Task(function=function, resource=resource, name=name, end=end))
```

mg-epsy replied:
This allows step functions to add services that are created outside the current serverless.yml. Earlier, task function was only taking the function as the argument. With this change, a new Task can be created with either function or a resource. Task was already taking resource but task function in stepfunctions.py was not exposing the resources. 

Usage is in PR https://github.com/epsyhealth/gdpr-user-data-extraction-service/blob/2cd5e25f81d4f278755005b12eeb14305180bed5/serverless.yml.py#L97-L107

whisller replied:
What I am saying is that: 
```python
def task(self, function=None, resource=None, name=None, end: Optional[bool] = None):
        if function is not None and resource is None:
            return self.definition.add(Task(function, end=end))
        elif function is None and resource is not None:
            return self.definition.add(Task(resource=resource, name=name, end=end))
        else:
            raise ValueError("Either 'function' or 'resource' must be provided, but not both.")
```

and

```python
    def task(self, function=None, resource=None, name=None, end: Optional[bool] = None):
        return self.definition.add(Task(function=function, resource=resource, name=name, end=end))
```

produces same output. 

Or I missed something?

mg-epsy replied:
I see your question now. sorry, i didnt understand before. :-)
Yes, they are the same but as this is a public repo/tool that could be used by others. So, I wanted the usage (combination of right args) to be more intuitive/clear to the end user. 
Thats all.

whisller replied:
@mg-epsy how is that cleaner? Main question is, is your six lines doing same as one line from my snippet?

mg-epsy replied:
yes, changed to 1 liner.

mg-epsy committed:
PR comment change

mg-epsy committed:
parallel does not have end property

Removing `End` since Parallel does not have `End` property.
https://www.serverless.com/plugins/serverless-step-functions#parallel

```
stepFunctions:
  stateMachines:
    yourParallelMachine:
      definition:
        Comment: "An example of the Amazon States Language using a parallel state to execute two branches at the same time."
        StartAt: Parallel
        States:
          Parallel:
            Type: Parallel
            Next: Final State
            Branches:
            - StartAt: Wait 20s
              States:
                Wait 20s:
                  Type: Wait
                  Seconds: 20
                  End: true
            - StartAt: Pass
              States:
                Pass:
                  Type: Pass
                  Next: Wait 10s
                Wait 10s:
                  Type: Wait
                  Seconds: 10
                  End: true
          Final State:
            Type: Pass
            End: true
```

mg-epsy commented on serverless/aws/features/stepfunctions.py:
```diff
@@ -237,11 +237,10 @@ def to_yaml(cls, dumper, data):
 class Parallel(Stage):
     yaml_tag = "!Parallel"
 
```
Removing `End` since Parallel throws error with `End` property during sls deploy
https://www.serverless.com/plugins/serverless-step-functions#parallel

```
stepFunctions:
  stateMachines:
    yourParallelMachine:
      definition:
        Comment: "An example of the Amazon States Language using a parallel state to execute two branches at the same time."
        StartAt: Parallel
        States:
          Parallel:
            Type: Parallel
            Next: Final State
            Branches:
            - StartAt: Wait 20s
              States:
                Wait 20s:
                  Type: Wait
                  Seconds: 20
                  End: true
            - StartAt: Pass
              States:
                Pass:
                  Type: Pass
                  Next: Wait 10s
                Wait 10s:
                  Type: Wait
                  Seconds: 10
                  End: true
          Final State:
            Type: Pass
            End: true
```

whisller replied:
As far as I can see in AWS documentation it does have `End` https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-parallel-state.html

mg-epsy replied:
Yes, I see that. However, here is what I am seeing.


- Without the above changes, genereated serverless.yml configuration looks as below with `End` attribute.
- Without the above changes, with the `End` attribute, when I deploy with ... sls deploy ... , I get the following error.
- Based on the error (shown below), I looked at the [serverless-step-function plugin's parallel configuration.](https://www.serverless.com/plugins/serverless-step-functions#parallel) which does not have an attribute called `End` for parallel type.
- Based on above obseverations, I made the changes. With the changes, serverless.yml generated can deploy successfully.

```
# Error with ... sls deploy ....

✖ Stack user-data-export-service-dev failed to deploy (20s)
Environment: darwin, node 19.6.0, framework 3.30.1 (local) 3.21.0v (global), plugin 6.2.3, SDK 4.3.2
Credentials: Local, environment variables
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
✕ State machine "user-data-export" definition is invalid:
SCHEMA_VALIDATION_FAILED: /States/ExportUserDataFromAllServices/End is invalid. must be equal to one of the allowed values
```

```
###
# Generated serverless.yml's relevant StepFunction configuration that causes the above error.
###
....
...
definition:
        StartAt: ExportUserDataFromAllServices
        Comment: export user data
        States:
          ExportUserDataFromAllServices:
            Type: Parallel
            Branches:
            - StartAt: Userservice
              States:
                Userservice:
                  Type: Task
                  ....
            - StartAt: VideoConverterService
              States:
                VideoConverterService:
                  Type: Task
                  ...
            End: false       # Error at this line
            ResultPath: null
            Next: PostExportProcess
            Catch:
            - ErrorEquals:
              - States.ALL
              Next: StateMachineErrorFallback
          PostExportProcess:
            Type: Task
            Resource:
              Fn::GetAtt:
              - PostExportProcessLambdaFunction
              - Arn
            End: true
            Catch:
            - ErrorEquals:
              - States.ALL
              Next: StateMachineErrorFallback
          StateMachineErrorFallback:
            Type: Pass
            Result: Error in state machine
            End: true
   
      
````
